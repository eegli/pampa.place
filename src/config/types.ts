import {
  BBox,
  Feature,
  FeatureCollection,
  MultiPolygon,
  Polygon,
} from '@turf/helpers';

export type MarkerConfig = {
  colors: Record<string, string>;
  svgMarker: {
    path: string;
    anchor: [number, number];
  };
};

export type GoogleConfig = {
  svRequest: {
    radius: number;
  }; // google.maps.StreetViewLocationRequest;
  streetview: google.maps.StreetViewPanoramaOptions;
  map: google.maps.MapOptions;
};

export type GameConfig = {
  maxPlayers: number;
  rounds: [number, number, number];
  roundsDefault: number;
  timeLimits: [number, number, number, number];
  timeLimitsDefault: number;
};

export type LatLngLiteral = { lat: number; lng: number };

// All maps eventually only need the name property
export type Properties = {
  name: string;
};

export type MapCollection = FeatureCollection<
  Polygon | MultiPolygon,
  Properties
>;
// Specific properties of the map data set of the EU
export type EUMapCollection = FeatureCollection<
  Polygon | MultiPolygon,
  {
    CNTR_CODE: string;
    NAME_LATN: string;
    NUTS_ID: string;
    FID: string;
  }
>;

// Full, computed map data
export type Maps<T = Properties> = Record<
  string,
  {
    computed: {
      // Area in km^2
      area: number;
      // Not needed for now
      center: LatLngLiteral;
      // Used to generate a random point
      bb: BBox;
      // Poly bounding box: SW SE NE NW
      bbLiteral: Record<'NE' | 'SE' | 'SW' | 'NW', LatLngLiteral>;
    };
    // Base can be used directly with google maps
    geo: Feature<Polygon | MultiPolygon, T>;
  }
>;

export type MapIdCollection = {
  name: string;
  id: string;
}[];
