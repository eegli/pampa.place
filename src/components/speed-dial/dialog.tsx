import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from '@mui/material';

export type DialogProps = {
  open: boolean;
  title: string;
  callback: () => void | (() => Promise<void>);
  callbackTitle: string;
  cancel: () => void;
  message?: string;
};

const ConfirmationDialog = ({
  open,
  title,
  callback,
  cancel,
  message,
  callbackTitle,
}: DialogProps) => {
  // Needed for smooth unmount transitions
  if (!open) {
    return <div />;
  }
  return (
    <div>
      <Dialog
        open={open}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
        PaperProps={{ elevation: 2 }}
      >
        <DialogTitle id="alert-dialog-title">{title}</DialogTitle>
        {message && (
          <DialogContent>
            <DialogContentText id="alert-dialog-description">
              {message}
            </DialogContentText>
          </DialogContent>
        )}
        <DialogActions>
          <Button onClick={cancel}>Cancel</Button>
          <Button onClick={callback} autoFocus>
            {callbackTitle}
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  );
};

export default ConfirmationDialog;
