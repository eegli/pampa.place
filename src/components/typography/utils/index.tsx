import Box from '@mui/material/Box';

export const em = (txt: string, color: 'g' | 'v') => (
  <Box
    component="span"
    color={color === 'g' ? 'success.main' : 'secondary.main'}
  >
    <b>{txt}</b>
  </Box>
);
