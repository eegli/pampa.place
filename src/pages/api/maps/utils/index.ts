export {apiData as data} from '@/config/maps';
import {NextApiHandler} from 'next';

type ErrorResponse = {
  error: string;
};

type DataResponse<T> = T extends unknown[]
  ? {
      data: T;
    }
  : {
      data: T[];
    };

export type ApiJSONHandler<T = unknown> = NextApiHandler<
  DataResponse<T> | ErrorResponse
>;
export type ApiTextHandler = NextApiHandler<string> | ErrorResponse;

export function nthQuery<T>(arg: T | T[], index = 0): T | undefined {
  return Array.isArray(arg) ? arg[index] : arg;
}
