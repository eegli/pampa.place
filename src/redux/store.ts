import {Constants} from '@/config/constants';
import {
  configureStore,
  Middleware,
  StateFromReducersMapObject,
} from '@reduxjs/toolkit';
import {createLogger} from 'redux-logger';
import app from './app';
import game from './game';
import position from './position';

const reducer = {
  position: position.reducer,
  app: app.reducer,
  game: game.reducer,
};

const logger = createLogger();

const windowStorage: Middleware<{}, RootState> = state => next => action => {
  if (action.type.includes('setApiKey') && typeof action.payload === 'string') {
    window.sessionStorage.setItem(Constants.SESSION_APIKEY_KEY, action.payload);
  }
  if (action.type.includes('setTheme')) {
    window.localStorage.setItem(Constants.THEME_KEY, action.payload);
  }
  return next(action);
};

const isDev = process.env.NODE_ENV === 'development';
const devMiddleware = [logger, windowStorage];
const prodMiddleware = [windowStorage];

export const initialStates = {
  position: position.reducer(undefined, {type: '@@INIT'}),
  app: app.reducer(undefined, {type: '@@INIT'}),
  game: game.reducer(undefined, {type: '@@INIT'}),
};

export function createStore(preloadedState?: RootState) {
  return configureStore({
    reducer,
    preloadedState,
    devTools: false,
    middleware: getDefault =>
      isDev
        ? getDefault().concat(devMiddleware)
        : getDefault().concat(prodMiddleware),
  });
}

export const store = createStore();

export type AppDispatch = typeof store.dispatch;
export type RootState = StateFromReducersMapObject<typeof reducer>;
