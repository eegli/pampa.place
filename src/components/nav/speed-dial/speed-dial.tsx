import {Dialog, DialogProps} from '@/components/feedback/dialog';
import {setTheme} from '@/redux/app';
import {resetRound} from '@/redux/game';
import {useAppDispatch, useAppSelector} from '@/redux/hooks';
import DarkModeIcon from '@mui/icons-material/DarkMode';
import HomeIcon from '@mui/icons-material/Home';
import LightModeIcon from '@mui/icons-material/LightMode';
import RestartIcon from '@mui/icons-material/RestartAlt';
import ShortcutIcon from '@mui/icons-material/Shortcut';
import Box from '@mui/material/Box';
import SpeedDial from '@mui/material/SpeedDial';
import SpeedDialAction from '@mui/material/SpeedDialAction';
import {useRouter} from 'next/router';
import {useState} from 'react';

type DialogStateProps = Omit<DialogProps, 'onCancelCallback'>;

export const SpeedDialNav = () => {
  const dispatch = useAppDispatch();
  const router = useRouter();
  const activeTheme = useAppSelector(s => s.app.theme);
  const [dialog, setDialog] = useState<DialogStateProps | undefined>(undefined);

  const dialogProps: DialogProps | undefined = dialog && {
    ...dialog,
    onCancelCallback: () => setDialog(undefined),
  };

  function handleToggleTheme() {
    dispatch(setTheme(activeTheme === 'light' ? 'dark' : 'light'));
  }

  return (
    <>
      <Box
        id="speed-dial-menu-overlay"
        zIndex={100}
        position="fixed"
        bottom={30}
        left={30}
        sx={{
          height: 330,
          transform: 'translateZ(0px)',
          flexGrow: 1,
        }}
      >
        <SpeedDial
          ariaLabel="speed-dial-menu"
          sx={{position: 'absolute', bottom: 16, left: 16, zIndex: 100}}
          icon={<ShortcutIcon />}
        >
          <SpeedDialAction
            key="restart"
            aria-label="restart"
            icon={<RestartIcon />}
            tooltipTitle="Restart round"
            tooltipPlacement="right"
            onClick={() => {
              setDialog({
                title: 'Restart round?',
                infoMessage:
                  'This will reset the current round progress for all players. The first player will start the current round again in a new location.',
                confirmMessage: 'Restart round',
                onConfirmCallback: function () {
                  setDialog(undefined);
                  dispatch(resetRound());
                },
              });
            }}
          />
          <SpeedDialAction
            key="home"
            aria-label="home"
            icon={<HomeIcon />}
            tooltipTitle="Home"
            tooltipPlacement="right"
            onClick={() => {
              setDialog({
                title: 'Abort the game and return home?',
                infoMessage: 'This will reset the current game',
                confirmMessage: 'Abort game',
                onConfirmCallback: async function () {
                  setDialog(undefined);
                  await router.push('/');
                },
              });
            }}
          />
          <SpeedDialAction
            key="theme"
            aria-label="toggle-theme"
            icon={
              activeTheme === 'light' ? <DarkModeIcon /> : <LightModeIcon />
            }
            tooltipTitle={activeTheme === 'light' ? 'Lights off' : 'Lights on'}
            tooltipPlacement="right"
            onClick={handleToggleTheme}
          />
        </SpeedDial>
      </Box>

      {dialogProps && <Dialog {...dialogProps} />}
    </>
  );
};
