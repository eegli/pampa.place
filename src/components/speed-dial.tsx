import DarkModeIcon from '@mui/icons-material/DarkMode';
import FlagIcon from '@mui/icons-material/Flag';
import HomeIcon from '@mui/icons-material/Home';
import LightModeIcon from '@mui/icons-material/LightMode';
import RestartAltIcon from '@mui/icons-material/RestartAlt';
import ShortcutIcon from '@mui/icons-material/Shortcut';
import Box from '@mui/material/Box';
import SpeedDial from '@mui/material/SpeedDial';
import SpeedDialAction from '@mui/material/SpeedDialAction';
import { useRouter } from 'next/router';
import { useState } from 'react';
import { useAppDispatch, useAppSelector } from '../redux/hooks';
import { toggleTheme } from '../redux/slices/app';
import { RootState } from '../redux/store';
import Dialog, { DialogProps } from './dialog';

export default function SpeedDialTooltipOpen() {
  const [dialog, setDialog] = useState<DialogProps>({} as DialogProps);

  const router = useRouter();

  const homeDialog: DialogProps = {
    title: 'Abort the game and return home?',
    callbackTitle: 'Abort game',
    callback: function () {
      router.push('/');
    },
    open: true,
    cancel: function () {
      setDialog({ ...dialog, open: false });
    },
  };

  const restartDialog: DialogProps = {
    title: 'Restart the game?',
    callbackTitle: 'Restart',
    callback: function () {
      router.push('/');
    },
    open: true,
    cancel: function () {
      setDialog({ ...dialog, open: false });
    },
  };

  // TODO
  const giveUpDialog: DialogProps = {
    title: 'Give up and skip round?',
    callbackTitle: 'Give up',
    callback: function () {
      router.push('/');
    },
    open: true,
    cancel: function () {
      setDialog({ ...dialog, open: false });
    },
  };

  const activeTheme = useAppSelector((s: RootState) => s.app.theme);
  const dispatch = useAppDispatch();

  return (
    <>
      <Box
        id="speed-dial-menu-overlay"
        zIndex={100}
        position="fixed"
        bottom={30}
        left={30}
        sx={{
          height: 330,
          transform: 'translateZ(0px)',
          flexGrow: 1,
        }}
      >
        <SpeedDial
          ariaLabel="SpeedDial tooltip example"
          sx={{ position: 'absolute', bottom: 16, left: 16, zIndex: 100 }}
          icon={<ShortcutIcon />}
        >
          <SpeedDialAction
            key="Give up"
            icon={<FlagIcon />}
            tooltipTitle="Give up"
            tooltipPlacement="right"
            onClick={() => setDialog(giveUpDialog)}
          />
          <SpeedDialAction
            key="Restart"
            icon={<RestartAltIcon />}
            tooltipTitle="Restart"
            tooltipPlacement="right"
            onClick={() => setDialog(restartDialog)}
          />
          <SpeedDialAction
            key="Home"
            icon={<HomeIcon />}
            tooltipTitle="Home"
            tooltipPlacement="right"
            onClick={() => setDialog(homeDialog)}
          />
          <SpeedDialAction
            key="Mode"
            icon={
              activeTheme === 'light' ? <DarkModeIcon /> : <LightModeIcon />
            }
            tooltipTitle={activeTheme === 'light' ? 'Lights off' : 'Lights on'}
            tooltipPlacement="right"
            onClick={() => dispatch(toggleTheme())}
          />
        </SpeedDial>
      </Box>
      <Dialog {...dialog} />
    </>
  );
}
