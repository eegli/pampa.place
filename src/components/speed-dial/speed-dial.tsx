import {toggleTheme} from '@/redux/app/app.slice';
import {resetRound} from '@/redux/game/game.slice';
import {useAppDispatch, useAppSelector} from '@/redux/redux.hooks';
import {RootState} from '@/redux/redux.store';
import DarkModeIcon from '@mui/icons-material/DarkMode';
import FlagIcon from '@mui/icons-material/Flag';
import HomeIcon from '@mui/icons-material/Home';
import LightModeIcon from '@mui/icons-material/LightMode';
import RestartAltIcon from '@mui/icons-material/RestartAlt';
import ShortcutIcon from '@mui/icons-material/Shortcut';
import Box from '@mui/material/Box';
import SpeedDial from '@mui/material/SpeedDial';
import SpeedDialAction from '@mui/material/SpeedDialAction';
import {useRouter} from 'next/router';
import {useState} from 'react';
import Dialog, {DialogProps} from './dialog';

export default function SpeedDialTooltipOpen() {
  const [dialog, setDialog] = useState<DialogProps>({} as DialogProps);

  const router = useRouter();

  const activeTheme = useAppSelector((s: RootState) => s.app.theme);
  const dispatch = useAppDispatch();

  return (
    <>
      <Box
        id="speed-dial-menu-overlay"
        zIndex={100}
        position="fixed"
        bottom={30}
        left={30}
        sx={{
          height: 330,
          transform: 'translateZ(0px)',
          flexGrow: 1,
        }}
      >
        <SpeedDial
          ariaLabel="SpeedDial tooltip example"
          sx={{position: 'absolute', bottom: 16, left: 16, zIndex: 100}}
          icon={<ShortcutIcon />}
        >
          <SpeedDialAction
            key="Give up"
            icon={<FlagIcon />}
            tooltipTitle="Give up"
            tooltipPlacement="right"
            onClick={() =>
              setDialog({
                title: 'Give up and skip round?',
                callbackTitle: 'Give up',
                callback: function () {
                  router.push('/');
                },
                open: true,
                cancel: function () {
                  setDialog({...dialog, open: false});
                },
              })
            }
          />
          <SpeedDialAction
            key="Restart"
            icon={<RestartAltIcon />}
            tooltipTitle="Restart round"
            tooltipPlacement="right"
            onClick={() =>
              setDialog({
                title: 'Restart round?',
                message:
                  'This will reset the current round progress for all players. The first player will start the current round again in a new location.',
                callbackTitle: 'Restart round',
                callback: function () {
                  dispatch(resetRound());
                  setDialog({...dialog, open: false});
                },
                open: true,
                cancel: function () {
                  setDialog({...dialog, open: false});
                },
              })
            }
          />
          <SpeedDialAction
            key="Home"
            icon={<HomeIcon />}
            tooltipTitle="Home"
            tooltipPlacement="right"
            onClick={() =>
              setDialog({
                title: 'Abort the game and return home?',
                callbackTitle: 'Abort game',
                callback: function () {
                  router.push('/');
                },
                open: true,
                cancel: function () {
                  setDialog({...dialog, open: false});
                },
              })
            }
          />
          <SpeedDialAction
            key="Mode"
            icon={
              activeTheme === 'light' ? <DarkModeIcon /> : <LightModeIcon />
            }
            tooltipTitle={activeTheme === 'light' ? 'Lights off' : 'Lights on'}
            tooltipPlacement="right"
            onClick={() => dispatch(toggleTheme())}
          />
        </SpeedDial>
      </Box>
      <Dialog {...dialog} />
    </>
  );
}
