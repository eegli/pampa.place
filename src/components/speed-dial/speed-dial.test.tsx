import {
  createMockState,
  createMockStore,
  fireEvent,
  render,
  screen,
} from '@/tests/utils';
import {RootState} from '../../redux/store';
import {DeepPartial, PickActual} from '../../utils/types';
import {SpeedDialNav} from './speed-dial';

const mockPush = jest.fn().mockResolvedValue(true);

jest.spyOn(require('next/router'), 'useRouter').mockReturnValue({
  push: mockPush,
});

beforeEach(() => {
  jest.clearAllMocks();
});

function hoverSpeedDial() {
  const speedDial = screen.getByRole('button', {name: 'speed-dial-menu'});
  fireEvent.mouseOver(speedDial);
}

function expectDialogToBeInDocument() {
  expect(screen.getByRole('dialog')).toBeInTheDocument();
}

function expectDialogToBeGone() {
  expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
}

describe('Speed dial', () => {
  it('renders 3 items on hover', () => {
    render(<SpeedDialNav />);
    expect(screen.getAllByRole('menuitem')).toHaveLength(3);
  });
  it('has theme toggle', () => {
    const state = createMockState({app: {theme: 'dark'}});
    const store = createMockStore(state);
    const {unmount} = render(<SpeedDialNav />, store);
    let toggleButton = screen.getByRole('menuitem', {name: /lights/i});
    fireEvent.click(toggleButton);
    expect(store.getState().app.theme).toBe('light');
    unmount();
    render(<SpeedDialNav />, store);
    toggleButton = screen.getByRole('menuitem', {name: /lights/i});
    fireEvent.click(toggleButton);
    expect(store.getState().app.theme).toBe('dark');
  });
  it('has home navigation', () => {
    render(<SpeedDialNav />);
    hoverSpeedDial();
    const homeButton = screen.getByRole('menuitem', {name: /home/i});
    fireEvent.click(homeButton);
    expectDialogToBeInDocument();
    const cancelButton = screen.getByRole('button', {name: /cancel/i});
    fireEvent.click(cancelButton);
    expectDialogToBeGone();
    fireEvent.click(homeButton);
    const abortGameButton = screen.getByRole('button', {name: /abort game/i});
    fireEvent.click(abortGameButton);
    expectDialogToBeGone();
    expect(mockPush).toHaveBeenCalledTimes(1);
    expect(mockPush).toHaveBeenCalledWith('/');
  });
  it('can restart round', () => {
    // Second round, second player's turn
    const store = createMockStore(
      createMockState({
        game: {
          players: ['player2', 'player1'],
          rounds: {
            total: 2,
            current: 2,
            progress: 1,
          },
          scores: [
            [
              {name: 'player2', score: 0},
              {name: 'player1', score: 0},
            ],
            [{name: 'player2', score: 0}],
          ],
        },
      })
    );
    render(<SpeedDialNav />, store);
    hoverSpeedDial();
    const restartButton = screen.getByRole('menuitem', {name: /restart/i});
    fireEvent.click(restartButton);
    expectDialogToBeInDocument();
    const cancelButton = screen.getByRole('button', {name: /cancel/i});
    fireEvent.click(cancelButton);
    expectDialogToBeGone();
    fireEvent.click(restartButton);
    const restartRoundButton = screen.getByRole('button', {
      name: /restart round/i,
    });
    fireEvent.click(restartRoundButton);
    expectDialogToBeGone();
    // Second round is restarted, first player's turn again
    expect(store.getState().game).toMatchObject<
      DeepPartial<PickActual<RootState, 'game'>>
    >({
      players: ['player1', 'player2'],
      rounds: {
        total: 2,
        current: 2,
        progress: 0,
      },
      scores: [
        [
          {name: 'player2', score: 0},
          {name: 'player1', score: 0},
        ],
        [],
      ],
    });
  });
});
