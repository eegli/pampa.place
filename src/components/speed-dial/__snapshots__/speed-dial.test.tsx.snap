// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Speed dial can restart round: restart confirmation dialog 1`] = `
Object {
  "confirmMessage": "Restart round",
  "infoMessage": "This will reset the current round progress for all players. The first player will start the current round again in a new location.",
  "onCancelCallback": [Function],
  "onConfirmCallback": [Function],
  "title": "Restart round?",
}
`;

exports[`Speed dial has home navigation: home confirmation dialog 1`] = `
Object {
  "confirmMessage": "Abort game",
  "infoMessage": "This will reset the current game",
  "onCancelCallback": [Function],
  "onConfirmCallback": [Function],
  "title": "Abort the game and return home?",
}
`;
